dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+import numpy as np
+import numpy.linalg as npl
+import scipy.ndimage as ndimage
+from dipy.align import floating
+import dipy.align.vector_fields as vf
+from dipy.align.mattes import MattesBase, sample_domain_regular
+from dipy.core.optimize import Optimizer
+from dipy.align.imwarp import (get_direction_and_spacings, ScaleSpace)
+from dipy.align.scalespace import IsotropicScaleSpace
+from dipy.core.optimize import SCIPY_LESS_0_12
+
+
+class MattesMIMetric(MattesBase):
+    def __init__(self, nbins=32, sampling_proportion=None):
+        r""" Initializes an instance of the Mattes MI metric
+
+        This class implements the methods required by Optimizer to drive the
+        registration process by making calls to the low level methods defined
+        in MattesBase.
+
+        Parameters
+        ----------
+        nbins : int, optional
+            the number of bins to be used for computing the intensity
+            histograms. The default is 32.
+        sampling_proportion : int in (0,100], optional
+            the percentage of voxels to be used for estimating the (joint and
+            marginal) intensity histograms. If None, dense sampling is used
+            (see notes regarding the difference between sparse and dense
+            sampling). The default is None.
+
+        Notes
+        -----
+        Since we use linear interpolation, images are not, in general,
+        differentiable at exact voxel coordinates, but they are differentiable
+        between voxel coordinates. When using sparse sampling, selected voxels
+        are slightly moved by adding a small random displacement within one
+        voxel to prevent sampling points to be located exactly at voxel
+        coordinates. When using dense sampling, this random displacement is
+        not applied.
+        """
+        super(MattesMIMetric, self).__init__(nbins)
+        self.sampling_proportion = sampling_proportion
+
+    def setup(self, transform, static, moving, static_grid2space=None,
```

Comment:
Sure!, 
`MattesBase`: computes the marginal and joint distributions of intensities of a pair of images, according to Mattes, which uses splines to make the histograms differentiable. It also computes the gradient of the joint histogram w.r.t. the parameters of a given transform. It also contains methods to compute the Mutual Information from the joint distribution and its gradient, maybe this is what seems odd: putting the particular metric MI inside MattesBase. Perhaps it would have been more clear to define a new class MattesMI extending MattesBase which defines MI and its gradient. Later we may have defined MattesCC, MattesCR and so on.

`MattesMIMetric`: it is a Python interface to the MattesBase cython class, which is used to compute the value and gradient the way Optimizer needs them. That is, given a set of transform parameters, it will use MattesBase methods to compute the value and gradient evaluated at the given parameters.

`AffineRegistration`: it runs the multi-resolution registration, putting all the pieces together. It needs to create the scale space of the images and run the multi-resolution registration by using the Metric and the Optimizer at each level of the Gaussian pyramid. At each level, it will setup the metric to compute value and gradient of the metric with the input images with different levels of smoothing. Then, when the Optimizer calls the metric and gradient from the metric, the metric is already configured to use the right images at the current resolution.